// simple-git/src/lib/tasks/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logTask = exports.parseLogOptions = void 0;
const parse_list_log_summary_1 = require("../parsers/parse-list-log-summary");
const utils_1 = require("../utils");
const excludeOptions = new Set([
    '--pretty',
    'max-count',
    'maxCount',
    'n',
    'file',
    'format',
    'from',
    'to',
    'splitter',
    'symmetric',
    'multiLine',
    'strictDate',
]);
function prettyFormat(format, splitter) {
    const fields = [];
    const formatStr = [];
    Object.entries(format).forEach(([field, format]) => {
        fields.push(field);
        formatStr.push(format);
    });
    return [
        fields, formatStr.join(splitter)
    ];
}
function userOptions(options) {
    return Object.fromEntries(Object.entries(options).filter(([key]) => !excludeOptions.has(key)));
}
function parseLogOptions(opt = {}, customArgs = []) {
    const splitter = opt.splitter || parse_list_log_summary_1.SPLITTER;
    const format = opt.format || {
        hash: '%H',
        date: opt.strictDate === false ? '%ai' : '%aI',
        message: '%s',
        refs: '%D',
        body: opt.multiLine ? '%B' : '%b',
        author_name: '%aN',
        author_email: '%ae'
    };
    const [fields, formatStr] = prettyFormat(format, splitter);
    const suffix = [];
    const command = [
        `--pretty=format:${parse_list_log_summary_1.START_BOUNDARY}${formatStr}${parse_list_log_summary_1.COMMIT_BOUNDARY}`,
        ...customArgs,
    ];
    const maxCount = opt.n || opt['max-count'] || opt.maxCount;
    if (maxCount) {
        command.push(`--max-count=${maxCount}`);
    }
    if (opt.from && opt.to) {
        const rangeOperator = (opt.symmetric !== false) ? '...' : '..';
        suffix.push(`${opt.from}${rangeOperator}${opt.to}`);
    }
    if (opt.file) {
        suffix.push('--follow', opt.file);
    }
    utils_1.appendTaskOptions(userOptions(opt), command);
    return {
        fields,
        splitter,
        commands: [
            ...command,
            ...suffix,
        ],
    };
}
exports.parseLogOptions = parseLogOptions;
function logTask(splitter, fields, customArgs) {
    return {
        commands: ['log', ...customArgs],
        format: 'utf-8',
        parser: parse_list_log_summary_1.createListLogSummaryParser(splitter, fields),
    };
}
exports.logTask = logTask;
